### PR TITLE
Rework number parser and angle support

### PR DESCRIPTION
**Pull request by Efnilite**
### Description
Reworks number parser. Adds support for angles. When using `x degrees`, it is just syntactic sugar for `x`. When using `x radians`, converts `x` to degrees.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]
This should be baked into the number parser. That way it can be used in contexts where only literals are accepted

**Review by cheeezburga — APPROVED** [review]
Only had a quick look through but looks good! The only thing I'd say is that the test could maybe cover some edge cases too (e.g. NaN, infinity, invalid values like strings, etc.)?

File: src/main/java/ch/njol/skript/classes/data/JavaClasses.java [hunk 1]
```diff
@@ -31,15 +13,20 @@
...
+import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class JavaClasses {
 
 	public static final int VARIABLENAME_NUMBERACCURACY = 8;
-	public static final Pattern INTEGER_PATTERN = Pattern.compile("-?[0-9]+");
-	public static final Pattern NUMBER_PATTERN = Pattern.compile("-?[0-9]+(?>\\.[0-9]+)?%?");
+	public static final Pattern INTEGER_PATTERN =
+		Pattern.compile("(?<num>-?[0-9]+)( (in )?((?<rad>rad(ian)?s?)|deg(ree)?s?))?");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/JavaClasses.java:27]:
```suggestion
		Pattern.compile("(?<num>-?[0-9]+)(?: (?:in )?(?:(?<rad>rad(?:ian)?s?)|deg(?:ree)?s?))?");
```
use non-capturing where possible

File: src/main/java/ch/njol/skript/classes/data/JavaClasses.java [hunk 2]
```diff
@@ -31,15 +13,20 @@
...
+	public static final Pattern DECIMAL_PATTERN =
+		Pattern.compile("(?<num>-?[0-9]+(?>\\.[0-9]+)?%?)( (in )?((?<rad>rad(ian)?s?)|deg(ree)?s?))?");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/JavaClasses.java:29]:
```suggestion
		Pattern.compile("(?<num>-?[0-9]+(?>\.[0-9]+)?%?)(?: (?:in )?(?:(?<rad>rad(?:ian)?s?)|deg(?:ree)?s?))?");
```

File: src/main/java/ch/njol/skript/classes/data/JavaClasses.java (hunk 1 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/classes/data/JavaClasses.java:27]: with the complexity given, would also be a good time to add documentation explaining the regex.

File: src/main/java/ch/njol/skript/classes/data/JavaClasses.java [hunk 3]
```diff
@@ -621,11 +239,483 @@ public void deserialize(String o, Fields f) {
...
+	 * Applies {@code stringToNumber} for parsing the number.
+	 * If the number is a percentage, it gets parsed as a double between 0-1.
+	 * Then tries to convert radians to degrees if the string contains a radian group.
+	 * If the string could be parsed, applies {@code converter} to convert the number to the desired type.
+	 * If the number is not valid by {@code isValid}, returns null.
+	 * </p>
+	 *
+	 * @param string The string with the possible number.
+	 * @param stringToNumber The function to parse the number, e.g. {@link Integer#parseInt(String)}.
+	 * @param fromPercentage A function that divides the value by 100.
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/JavaClasses.java:303]: is this really needed
how many implementations of dividing by 100 can there be?

File: src/main/java/ch/njol/skript/classes/data/JavaClasses.java [hunk 4]
```diff
@@ -621,11 +239,483 @@ public void deserialize(String o, Fields f) {
...
+	 * @param converter The function to convert the number to the desired type, e.g. {@link Number#intValue()}.
+	 * @param isValid The function to check if the number is valid, e.g. {@link Double#isNaN()}.
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/JavaClasses.java:305]: does this need to be done in the function or would it really harm anything to just return Number and check validity after?

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"180 degrees is pi # true",
+	"pi radians is 180 degrees # true"
+})
+@Since("INSERT VERSION")
+public class ExprAngle extends SimpleExpression<Number> {
+
+	static {
+		Skript.registerExpression(ExprAngle.class, Number.class, ExpressionType.SIMPLE,
+			"%number% [in ]deg[ree][s]",
+			"%number% [in ]rad[ian][s]");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:34]:
```suggestion
			"%number% [in] deg[ree][s]",
			"%number% [in] rad[ian][s]");
```
for ease of reading. Also, any reason multiple numbers aren't supported?

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	@Override
+	protected Number @Nullable [] get(Event event) {
+		Number number = angle.getSingle(event);
+
+		if (number == null)
+			return null;
+
+		if (isRadians) {
+			return new Double[]{Math.toDegrees(number.doubleValue())};
+		}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:58]: choose brackets or no brackets

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	public Class<? extends Number> getReturnType() {
+		return Number.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return angle.toString(event, debug) + (isRadians ? " degrees" : " radians");
+	}
+
+}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:78]:
```suggestion
}

```

File: src/test/skript/tests/misc/angle.sk [hunk 8]
```diff
@@ -0,0 +1,10 @@
+test "angle":
```
Comment (sovdeeth) [inline src/test/skript/tests/misc/angle.sk:1]: nans, infinities, nulls, etc need tests

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java (hunk 5 above)
Comment (Pikachu920) [reply src/main/java/ch/njol/skript/expressions/ExprAngle.java:34]: I think if we support multiple, it should require `in` to do so

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprAngle.java:34]: Agreed, that would help with `1, 200, and pi degrees` instances.
Though that case maybe makes me think singular only is good.

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Angle")
+@Description({
+	"Represents the passed number value in degrees.",
+	"If radians is specified, converts the passed value to degrees. This conversion may not be entirely accurate, " +
+		"due to floating point precision.",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:20]:
```suggestion
	"due to floating point precision.",
```

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprAngle.java:20]: not required but makes it easier to read imo

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		isRadians = matchedPattern % 2 != 0;
+		return true;
+	}
+
+	@Override
+	protected Number @Nullable [] get(Event event) {
+		Number[] numbers = angle.getAll(event);
+
+		if (numbers == null)
+			return null;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:56]: getAll cannot return null or contain nulls

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by sovdeeth — CHANGES_REQUESTED** [review]
tests failing :(

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprAngle.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+						Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
+		angle = (Expression<Number>) expressions[0];
+		isRadians = matchedPattern % 2 != 0;
+		return true;
+	}
+
+	@Override
+	protected Number @Nullable [] get(Event event) {
+		Number[] numbers = angle.getAll(event);
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprAngle.java:53]: Any reason you're using `getAll` here?

**Review by UnderscoreTud — APPROVED** [review]

**Review by Moderocky — APPROVED** [review]

**Comment by Moderocky** [comment]
Good PR. I wondered about supporting the symbols too (e.g. `90°`) but they're not very standardised so it's not essential.
